### PR TITLE
feat: use cluster autoscaler to improve deploy time

### LINKED ISSUE #118: Use cluster autoscaler to improve deploy time
For more information see https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler.

### PR DESCRIPTION
**Pull request by garethjevans**
fixes #118

### CODE REVIEW

**Comment by garethjevans** [comment]
NOTE: this change will require the cluster to be recreated, so we would need to think about this carefully.

**Comment by garethjevans** [comment]
A full build & destroy is 9mins rather than 28mins.

**Comment by hferentschik** [comment]
> this change will require the cluster to be recreated, so we would need to think about this carefully.

can you give some more context? I like the idea of going from 28  to 9 minutes, but what are the implications? You seem to imply that now on changes to the cluster config which otherwise would just change a cluster property, the cluster gets rebuild?

Is this - https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler - what you try to enable here?

Is it worth opening an issue and track what we are trying to achieve here as well as the pros and cons?

**Comment by garethjevans** [comment]
I created this based on feedback that the terraform was too slow.  This is the same logic for autoscaler that we use on the SaaS and it works well.  It's also the same information I sent to @hferentschik on the 16th April but noticed it hadn't got anywhere.

I thought opening a PR was the best place to discuss this.

**Comment by garethjevans** [comment]
Terraform cannot make this change "in place", so will delete and recreate the cluster - like I said, this needs to be thought about if this is behaviour we want to introduce

**Comment by hferentschik** [comment]
> I thought opening a PR was the best place to discuss this.

it's a great place for sure. An issue would imo be even better ;-)

**Comment by hferentschik** [comment]
> NOTE: this change will require the cluster to be recreated, so we would need to think about this carefully.

> Terraform cannot make this change "in place", so will delete and recreate the cluster - like I said, this needs to be thought about if this is behaviour we want to introduce

I think I misunderstood what you were trying to say. I thought you meant that once this change is made, changes to the cluster which atm just change/update things, will require a cluster rebuild.

However, reading your second comment, you mean that if we were to merge and release this change and a user tries to upgrade, the cluster would need to be rebuild. That raises the question of how we want to do versioning in general. Going from 28min to 9min deploy time is a huge win. I would love to get this in. 

What guarantees does one want to give during upgrades from one version of the module to the next? Semantic says:

> Given a version number MAJOR.MINOR.PATCH, increment the:

> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards compatible manner, and
> PATCH version when you make backwards compatible bug fixes.

So in its core semantic versioning is about API. The API of a Terraform module is its in- and outputs and the behaviour it creates, right? That does not really change here. So we would be talking PATCH level, maybe MINOR if we would also expose things like disk type from the main module. 

With Terraform another question comes into play. Does a version change need to take into account what happens when I have a created infrastructure and I update the module and re-apply? Does that need to be considered when determining the new version? There is https://www.terraform.io/docs/extend/best-practices/versioning.html, but it does not give a concrete answer to this question either and it is maybe more specific to providers. 

Should this change be marked as a MAJOR change, because an upgrade requires a recreation of the cluster? 

To a certain degree the problem is also a problem with Jenkins X. If there were a way to backup and restore a given cluster easily, one could always argue that in a case where the cluster needs rebuilding, one should backup, update cluster, restore. 

@garethjevans - does this make any sense? WDYT?

File: modules/cluster/variables.tf [hunk 1]
```diff
@@ -96,6 +124,12 @@ variable "node_preemptible" {
   default     = false
 }
 
+variable "node_disk_type" {
```
Comment (hferentschik) [inline modules/cluster/variables.tf:127]: should this not become a variable of the main module as well?

**Comment by hferentschik** [comment]
@garethjevans, I think this pull request needs a rebase as well

**Comment by hferentschik** [comment]
@garethjevans I will be trying to rebase this pull request myself since a merge commit has sneaked in here.

**Pull request merged**